Data/HashMap/Base.hs

```diff
@@ -934,6 +935,18 @@ difference a b = foldlWithKey' go empty a
                  _       -> m
 {-# INLINABLE difference #-}
 
+-- | /O(n*log m)/ Difference with a combining function. When two equal keys are
+-- encountered, the combining function is applied to the values of these keys.
+-- If it returns 'Nothing', the element is discarded (proper set difference). If
+-- it returns (@'Just' y@), the element is updated with a new value @y@.
+differenceWith :: (Eq k, Hashable k) => (v -> w -> Maybe v) -> HashMap k v -> HashMap k w -> HashMap k v
+differenceWith f a b = foldlWithKey' go empty a
+  where
+    go m k v = case lookup k b of
+                 Nothing -> insert k v m
+                 Just w  -> maybe m (\y -> insert k y m) (f v w)
```

Comment:
This is why this is tricky. If these functions take an existing value from the map, don't modify it, and stick it in a different map, then the value doesn't need to be evaluated (because the invariant is that values in a `Strict` map are already in WHNF). However, if the value comes from the user or is modified, then we need to evaluate it before inserting it.